src/Oscoin/Protocol.hs

```diff
@@ -0,0 +1,287 @@
+                 )
+import qualified Oscoin.Storage.Block.Abstract as BlockStore
+import           Oscoin.Storage.Block.Orphanage (Orphanage)
+import qualified Oscoin.Storage.Block.Orphanage as O
+import           Oscoin.Telemetry (fmtBlockHash, fmtValidationError)
+import           Oscoin.Telemetry.Logging (LogMsg, ftag)
+import           Oscoin.Time.Chrono as Chrono
+
+import           Control.Concurrent.Async (Async)
+import qualified Control.Concurrent.Async as Async
+import           Control.Concurrent.STM
+import qualified Data.List.NonEmpty as NonEmpty
+
+-- | A stateful handle over a 'Protocol'.
+data Handle c tx s = Handle
+    { hProtocol :: MVar (Protocol c tx s IO) }
+
+-- | This data structures incorporates all the different components and
+-- data structures in the system devoted to store \"things\". Certain
+-- components (like the block cache, for example) are not managed by the
+-- 'Protocol' directly but they are rather an implementation detail of
+-- whichever 'BlockStore' this manager encapsulates.
+data Protocol c tx s m = Protocol
+    { protoFullBlockStore :: FullBlockStoreAPI c tx s m
+    -- ^ The 'BlockStore', an opaque interface to a component which only
+    -- responsibility is to store and retrieve blocks.
+    , protoOrphanage      :: Orphanage c tx s
+    -- ^ The in-memory data structure which keeps orphans blocks and constructs
+    -- sub-chains suitable for fork selection.
+    , protoFetchNextBlock :: m (Maybe (Block c tx (Sealed c s)))
+    -- ^ A monadic action to fetch the next incoming block, if any. This allows
+    -- the 'Protocol' to not be coupled to a concrete queue (like a 'TBQueue')
+    -- thus allowing it to be run in a variety of different stacks.
+    , protoValidateFull   :: Validate c tx s
+    -- ^ A validation function to (fully) validate a block.
+    , protoConfig         :: Consensus.Config
+    }
+
+{------------------------------------------------------------------------------
+  Using the 'Protocol'
+------------------------------------------------------------------------------}
+
+data ShutdownPolicy =
+      ShutdownImmediately
+    -- ^ Don't wait for queued messages to be processed.
+    | GracefulShutdown
+    -- ^ Wait for the queued messages to all be processed.
+    deriving Show
+
+-- | Given the data necessary to initialise and build the 'Protocol', runs
+-- the input action and return the final result.
+withProtocol
+    :: forall c tx s a m.
+      m (Maybe (Block c tx (Sealed c s)))
+    -- ^ Monadic action to fetch the next block, if any.
+    -> Orphanage c tx s
+    -> Validate c tx s
+    -> FullBlockStoreAPI c tx s m
+    -> Consensus.Config
+    -> (Protocol c tx s m -> m a)
+    -> m a
+withProtocol fetchNextBlock o validateFull bs config action =
+    let protocol = Protocol bs o fetchNextBlock validateFull config
+    in action protocol
+
+-- | Initialises and runs the 'Protocol', via its stateful 'Handle'.
+runOscoinProtocol
```

Comment:
Yes, this is basically what I was thinking to do as well, great suggestion. This could be also used directly in the miner if we want to ensure the block is fully processed before we return from that function.